server/src/main/java/org/apache/druid/server/http/DataSourcesResource.java

```diff
@@ -278,13 +284,14 @@ public Response markSegmentsAsUnused(
       );
       return numUpdatedSegments;
     };
-    return performSegmentUpdate(dataSourceName, payload, operation);
+    return performSegmentUpdate(dataSourceName, payload, operation, true);
   }
 
   private Response performSegmentUpdate(
       String dataSourceName,
       MarkDataSourceSegmentsPayload payload,
-      SegmentUpdateOperation operation
+      SegmentUpdateOperation operation,
+      boolean validateDatasourceIsQueryable
```

Comment:
hmm, this parameter name is slightly confusing. The short circuiting logic below is an implementation detail for an optimization, so I don't think the callers need to be say whether to skip the validation checks on datasource or not.

How about calling it `boolean isMarkUsed` or something similar?

Reply:
Yeah, boolean args are always weird. How about we do this instead.

- Get rid of the `performSegmentUpdate` method, the one which accepts a `MarkDataSourceSegmentsPayload`.
- Create method `buildInvalidMarkSegmentsPayloadResponse`

Then, our methods would look like this, and it would be readily evident where we want to perform which validations:

```java
public Response markAsUsedNonOvershadowedSegments()
{
   final SegmentUpdateOperation operation = ...
   
   if (payload == null || !payload.isValid()) {
       return buildInvalidMarkSegmentsPayloadResponse(payload);
   } else {
       return performSegmentUpdate(dataSourceName, operation);
   }    
}
```

```java
public Response markSegmentsAsUnused()
{
   final SegmentUpdateOperation operation = ...
   
   if (payload == null || !payload.isValid()) {
       return buildInvalidMarkSegmentsPayloadResponse(payload);
   } else if (getQueryableDatasource(dataSourceName) == null) {
       return logAndCreateDataSourceNotFoundResponse(dataSourceName);
   } else {
       return performSegmentUpdate(dataSourceName, operation);
   }    
}

```

Reply:
Thanks, I took your suggestion here, but changed the order of the branches.